bigbone/src/main/kotlin/social/bigbone/MastodonClient.kt

```diff
@@ -698,22 +702,50 @@ private constructor(
         /**
          * Get the version string for this Mastodon instance.
          * @return a string corresponding to the version of this Mastodon instance
-         * @throws BigBoneRequestException if instance version can not be retrieved using any known method or API version
+         * @throws BigBoneClientInstantiationException if instance version cannot be retrieved using any known method or API version
          */
+        @Throws(BigBoneClientInstantiationException::class)
         private fun getInstanceVersion(): String {
-            try {
-                val serverInfoVersion = NodeInfoClient
-                    .retrieveServerInfo(instanceName)
-                    ?.software
-                    ?.takeIf { it.name == "mastodon" }
-                    ?.version
-                if (serverInfoVersion != null) return serverInfoVersion
-            } catch (_: BigBoneRequestException) {
+            return try {
+                getInstanceVersionViaServerInfo()
+            } catch (error: BigBoneClientInstantiationException) {
+                // fall back to retrieving from Mastodon API itself
+                try {
+                    getInstanceVersionViaApi()
+                } catch (instanceException: InstanceVersionRetrievalException) {
+                    throw BigBoneClientInstantiationException(
+                        message = "Failed to get instance version of $instanceName",
+                        cause = if (instanceException.cause == instanceException) {
+                            instanceException.initCause(error)
+                        } else {
+                            instanceException
+                        }
```

Comment:
This part is debatable and could be replaced with just the `instanceException`. I have added the cause to get as detailed a stack trace as possible.

This would be relevant in cases where we fail to get the instance version via server info and then also fail getting it via the API. It would then add the original server info exception as the cause of the `instanceException` we got when attempting to get the version via the Mastodon API.

Users _likely_ won’t be able to do much with the server info exception, so we could potentially remove this.


```suggestion
                        cause = instanceException
```

Reply:
As soon as we decide that the fallbacks are not worth it, I’m very happy to remove them.

But for the time being, this is likely the way that will get us the most insight (provided library users get in touch with us)